### PR TITLE
Add metadata to Fault filter.

### PR DESCRIPTION
**Pull request by rshriram**
Commit Message: Add metadata field to fault filter to aid logging. 
Additional Description: This metadata will be available as part of StreamInfo's dynamicMetadata().
Risk Level: Low
Testing: Unit tests
Docs Changes: Not Yet
Release Notes: Add metadata field to fault filter to aid logging.
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by yanavlasov** [comment]
Format error: https://dev.azure.com/cncf/envoy/_build/results?buildId=135371&view=logs&j=c7de00aa-b0fb-52ba-707b-da6afd918eda&t=363aefed-4bbc-548d-2491-7ca65630b0e9&l=75

**Comment by tyxia** [comment]
I think you need to add "//envoy/type/metadata/v3:pkg" to deps in its BUILD file.

**Review by tyxia — COMMENTED** [review]
I think you need to add "//envoy/type/metadata/v3:pkg" to deps in its BUILD file

Update: Sorry, this is wrong. My comment https://github.com/envoyproxy/envoy/pull/27003#pullrequestreview-1404391280 here is right way to fix.

**Review by adisuissa — DISMISSED** [review]
LGTM, thanks.
When this is implemented, I suggest adding docs that show how to use the metadata.
CI failure seems relevant. Can you PTAL?
/lgtm api

**Review by adisuissa — DISMISSED** [review]
LGTM, thanks!
/lgtm api

/assign-from @envoyproxy/senior-maintainers

**Review by zuercher — DISMISSED** [review]
lgtm

File: source/extensions/filters/http/fault/fault_filter.cc
Comment (tyxia) [inline source/extensions/filters/http/fault/fault_filter.cc:173]: Can we set metadata through `setDynamicMetadata` with namespace used in reverse DNS format

**Review by tyxia — COMMENTED** [review]
Thanks!

I left a few comments since I have looked at the test files already :)

File: api/envoy/extensions/filters/http/fault/v3/fault.proto
Comment (htuch) [inline api/envoy/extensions/filters/http/fault/v3/fault.proto:154]: Why is it dynamic metadata here but filter state in https://github.com/envoyproxy/envoy/pull/27003?

File: test/extensions/filters/http/fault/fault_filter_test.cc
Comment (rshriram) [reply test/extensions/filters/http/fault/fault_filter_test.cc:343]: It doesnt.. I dont fully understand why

File: test/extensions/filters/http/fault/fault_filter_test.cc
Comment (rshriram) [reply test/extensions/filters/http/fault/fault_filter_test.cc:648]: Nope. It is twice because there are two delay faults. The dynamic metadata is called for each

File: source/extensions/filters/http/fault/fault_filter.cc
Comment (rshriram) [reply source/extensions/filters/http/fault/fault_filter.cc:173]: How do we know the namespace? And do we know if the provided metadata has only one filter namespace ? It could have many filter related namespace + structs right?

File: api/envoy/extensions/filters/http/fault/v3/fault.proto
Comment (rshriram) [reply api/envoy/extensions/filters/http/fault/v3/fault.proto:154]: The dynamic metadata here is just purely protobuf struct. There is no stat or anything. This is equivalent to the per route metadata and cluster metadata. Its not conveying anything that is computed by this filter

File: source/extensions/filters/http/fault/fault_filter.cc
Comment (tyxia) [reply source/extensions/filters/http/fault/fault_filter.cc:173]: Here, it is `envoy.filters.http.fault`. `envoy.` namesapce is save for built-in filer. 
For non built-in filter, you can find example in google3's ilb/configuration_constants.h and how it is used to mutate dynamic metadata.

Metadata is a map of proto struct. The filter namespace is unique and fixed to each filter.  It is also why namespace (as key)  is needed: when multiple filters in filter chain, you need this key to find the associated metadata to use/mutate.

File: test/extensions/filters/http/fault/fault_filter_test.cc
Comment (tyxia) [reply test/extensions/filters/http/fault/fault_filter_test.cc:343]: hmm interesting. If order is important, we can use `InSequence`. 
Regardless, if test is consistent passing. I can take a look once it is merged, I just cannot imagine only last line of function words.

File: source/extensions/filters/http/fault/fault_filter.cc
Comment (rshriram) [reply source/extensions/filters/http/fault/fault_filter.cc:173]: Thats not what I am saying. If you look at the route metadata, it is plainly merged into dynamic metadata. It is not automagically suffixed under envoy.filters.http.router. The guidance is that users should specify the filter namespace in the metadata. And if you look at the tests, you will see that there is a filter namespace under which the protobuf struct is added. So I feel that there is no need for us to add an additional level into this dynamic metadata.

Comment (tyxia) [reply source/extensions/filters/http/fault/fault_filter.cc:173]: dynamic metadata in streamInfo can be shared with different filters in the chain. Underlying it is a map `map<string, google.protobuf.Struct>`

Think about the case here: `filter_a->filter_b ->filter_c ((you are here)) ->filter_d` what you are doing here is override all other filters's metadata with your filter_c's metadata. Instead, you should only update the data associated with your filter.
route metadata is different I think it is not used to be shared between different filters.

File: test/extensions/filters/http/fault/fault_filter_test.cc
Comment (tyxia) [reply test/extensions/filters/http/fault/fault_filter_test.cc:648]: I see. Then it will be same as `EXPECT_CALL()Times(2).WillRepeatedly()` ? I feel that will be clean but it is a cosmetic change up to you :)

**Review by htuch — DISMISSED** [review]
LGTM, thanks!

**Review by adisuissa — COMMENTED** [review]
/lgtm api
LGTM.

File: source/extensions/filters/http/fault/fault_filter.cc
Comment (rshriram) [reply source/extensions/filters/http/fault/fault_filter.cc:173]: Note that we are merging and not overwriting. So we will not overwrite any other filter's dynamic metadata as long as the namespaces they use do not collide. If two filters have the same metadata namespace, then the contents will be merged using usual protobuf struct merge semantics. This is the same principle as Route metadata.

**Review by htuch — APPROVED** [review]

**Review by htuch — APPROVED** [review]

**Review by htuch — APPROVED** [review]

**Review by htuch — APPROVED** [review]

File: source/extensions/filters/http/fault/fault_filter.cc
Comment (tyxia) [reply source/extensions/filters/http/fault/fault_filter.cc:173]: You are right that as long as the namespaces do not collide, it won't not overwrite.

However, I think it is wrong for the case that name space collide. It will be overwrite if. The specific test `DynamicMetadataTest` in stream_info_impl_test.cc can be used to show this overwrite behavior.

So basically, you are override the filter state of fault_filter every time `MergeFrom` is invoked.

Comment (htuch) [reply source/extensions/filters/http/fault/fault_filter.cc:173]: Actually,  think we do need to fix this. We should merge dynamic metadata here under the configured filter name as key, not on top of all existing dynamic metadata. That's broken IMHO as there is too much chance of conflict and it doesn't properly namespace.

**Comment by rshriram** [comment]
obsolete

**Pull request closed without merging**
